### ISSUE TITLE
buf_cstring can crash on "novalue"

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#3813]
This doesn't currently occur in practice, but I tripped over it while debugging something unrelated, and would like to not trip over it again!

`novalue` is declared here.  Note that it's declared `static const`:

https://github.com/cyrusimap/cyrus-imapd/blob/fc57d4b15dad2850843a84a188e276b0ba283fd2/imap/sync_support.c#L1476-L1486

`buf_cstring()` is defined here:

https://github.com/cyrusimap/cyrus-imapd/blob/fc57d4b15dad2850843a84a188e276b0ba283fd2/lib/util.c#L1104-L1110

Its intention is to always return a valid cstring, and convert the `struct buf` to contain one internally at the same time.  It takes a `const struct buf*` argument so that it can be used in const-correct code, and casts away the constness internally to implement the behaviour that makes sure the buffer is 0-terminated.  This is usually fine.

But objects declared `static const` (like `novalue`) are probably stored in a read-only segment, not on the stack or heap.  So it doesn't matter that we cast away the constness -- the memory is still read only, and it segfaults.

I think what I'd like to do is move `novalue` into lib/util.c, so that everything can use it if it wants (not just `apply_annotations()`), and then special-case buf_cstring to recognise this constant and just return `""` without trying to modify it internally.

**Comment by rsto** [comment]
Nice catch. Why not just make `novalue` a non-static variable? I don't see the point in the `static const` declaration. Also, special-casing `if buf == &novalue` on every call of `buf_cstring` doesn't look right to me.

Note: this also reminds me of updating the buf_cstring implementation to only add `\0` if the last character is non-zero. The current implementation asks for one more byte in the buffer each time `buf_cstring` is called. That could incur even the cost of a realloc!

**Comment by brong** [comment]
I'm also gonna push back against the special-casing - that stuff is weird.  Let's just unstatic it.

(the line blames me, but it was actually Greg who wrote it I think, old code now!)

And yeah, good call with the "buf_cstring and \0" magic.  I'd also suggest that maybe we always mmap at least one extra byte when we're mmaping, and allow buf_cstring to return the mmap base rather than mallocing.

**Comment by elliefm** [comment]
Cool, this is good feedback, thanks!  If we're happy to make `novalue` non-static, rather than working around it being static, I am too.

That blame on @brong is just from him reordering the file contents at some point, it is indeed much older.

**Comment by elliefm** [comment]
Regarding making buf_cstring skip adding the '\0' if it's already there -- looks like we had an implementation like that, way back when.  It kept a flag around for whether the internal buffer was known to be a valid C string.  And 894543bde1 found it was cheaper to just always add the '\0' than to check the flag.  Hmm.

I'm thinking of changing BUF_INITIALIZER like this:

```
-#define BUF_INITIALIZER { NULL, 0, 0, 0 }
+#define BUF_INITIALIZER { "", 0, 0, 0 } /* copy-on-write empty string */
```

I think everything should handle this cleanly -- it's the same as calling `buf_init_ro(&buf, "", 0)`, but usable as an initialiser.

But!  The whole point of doing so would be so that new empty bufs can be "empty C strings" without needing to malloc space for a '\0'.  And yet, in this case `buf->alloc` would be `0` (there is no allocation), and buf_ensure will always create a new allocation if there wasn't one yet.  So it doesn't actually help, unless we do something in buf_cstring to recognise the case and avoid the buf_ensure... but we already decided it was cheaper to just do the buf_ensure.

I thought about just wrapping the backdoor code in `if (buf->s[buf->len] != '\0') { ... }`, but that's not safe -- `buf->s[buf->len]` might be out of bounds, and by the time we've checked, we might as well have just `buf_ensure()`'d...

At this point, I'm seeing a lot of edge cases to be wary of, and not really seeing any clear benefit to tinkering with `struct buf`'s internals.  So I'm thinking I'll just remove the `static` keyword from that one `novalue` variable and be done with it.

We've talked recently about someday making a new thing like `struct buf` that's designed from the ground up to scratch the itches that `struct buf` creates -- maybe "empty strings shouldn't require a malloc" is one of those itchec that we can revisit if/when we do that.
